src/models/partial_tx.ts

```diff
@@ -0,0 +1,594 @@
+  type: string,
+  value: number,
+  tokenData: number,
+  address?: string,
+  data?: string,
+  timelock?: number,
+};
+
+// Transaction data to use with helpers
+type TxData = {
+  inputs: InputData[],
+  outputs: OutputData[],
+  tokens: string[],
+  weight?: number,
+  nonce?: number,
+  version?: number,
+  timestamp?: number,
+};
+
+/**
+ * Extended version of the Input class with extra data
+ * We need the extra data to calculate the balance of the PartialTx
+ */
+export class ProposalInput extends Input {
+  token: string;
+  tokenData: number;
+  value: number;
+  address: string;
+
+  constructor(
+    hash: string,
+    index: number,
+    value: number,
+    address: string,
+    {
+      token = HATHOR_TOKEN_CONFIG.uid,
+      tokenData = 0,
+    }: {
+      token?: string;
+      tokenData?: number,
+    } = {},
+  ) {
+    super(hash, index);
+    this.value = value;
+    this.tokenData = tokenData;
+    this.token = token;
+    this.address = address;
+  }
+
+  /**
+   * Return an object with the relevant input data
+   *
+   * @return {InputData}
+   * @memberof ProposalInput
+   * @inner
+   */
+  toData(): InputData {
+    const data: InputData = {
+      tx_id: this.hash,
+      index: this.index,
+      address: this.address,
+    };
+
+    return data;
+  }
+
+  isAuthority(): boolean {
+    return (this.tokenData & TOKEN_AUTHORITY_MASK) > 0;
+  }
+}
+
+/**
+ * Extended version of the Output class with extra data
+ * We need the extra data to calculate the token_data of the
+ * output on the final transaction and to track which outputs are change.
+ */
+export class ProposalOutput extends Output {
+  token: string;
+  isChange: boolean;
+
+  /**
+   * We do not set tokenData because the token array is not yet formed
+   */
+  constructor(
+    value: number,
+    script: Buffer,
+    {
+      isChange = false,
+      token = HATHOR_TOKEN_CONFIG.uid,
+      tokenData = 0,
+    }: {
+      token?: string,
+      isChange?: boolean,
+      tokenData?: number,
+    } = {},
+  ) {
+    super(value, script, { tokenData });
+    this.token = token;
+    this.isChange = isChange;
+  }
+
+  /**
+   * Set the value of the property tokenData
+   *
+   * @param {number} tokenData
+   */
+  setTokenData(tokenData: number) {
+    this.tokenData = tokenData;
+  }
+
+  /**
+   * Return an object with the relevant output data
+   *
+   * @param {number} tokenIndex Index of the token on the tokens array
+   * @param {Network} network Network used to generate addresses in
+   *
+   * @returns {OutputData}
+   *
+   * @throws {UnsupportedScriptError} Script must be P2SH or P2PKH
+   * @memberof ProposalOutput
+   * @inner
+   */
+  toData(tokenIndex: number, network: Network): OutputData {
+    const script = this.parseScript(network);
+    if (!(script instanceof P2PKH || script instanceof P2SH)) {
+      throw new UnsupportedScriptError('Unsupported script type');
+    }
+
+    const authority = this.tokenData & TOKEN_AUTHORITY_MASK;
```

Comment:
This class already have an `isAuthority` since it inherits from the `Output` model,
I just think that below where we update the `tokenData` it would be a little confusing:
```js
this.setTokenData((this.isAuthority() ? TOKEN_AUTHORITY_MASK : 0) | tokenIndex)
```

Reply:
Ahhh makes sense, it's fine